subprojects/enterprise-operations/src/main/java/org/gradle/operations/dependencies/transforms/SnapshotTransformInputsBuildOperationType.java

```diff
@@ -0,0 +1,98 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.operations.dependencies.transforms;
+
+import org.gradle.api.internal.tasks.SnapshotTaskInputsBuildOperationType;
+import org.gradle.internal.operations.BuildOperationType;
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents the computation of capturing the before execution state and resolving the caching state for transforms.
+ * <p>
+ * Must occur as a child of {@link org.gradle.operations.execution.ExecuteWorkBuildOperationType}.
+ *
+ * @since 8.3
+ */
+public final class SnapshotTransformInputsBuildOperationType implements BuildOperationType<SnapshotTransformInputsBuildOperationType.Details, SnapshotTransformInputsBuildOperationType.Result> {
+    public interface Details {
+    }
+
+    /**
+     * The hashes of the inputs.
+     * <p>
+     * If the inputs were not snapshotted, all fields are null.
+     */
+    public interface Result {
+
+        /**
+         * The overall hash value for the inputs.
+         * <p>
+         * Null if the overall key was not calculated because the inputs were invalid.
+         */
+        @Nullable
+        byte[] getHashBytes();
+
+        /**
+         * The hash of the classloader that loaded the transform action implementation.
+         * <p>
+         * Null if the classloader is not managed by Gradle.
+         */
+        @Nullable
+        byte[] getClassLoaderHashBytes();
+
+        /**
+         * The class name of the transform's action.
+         */
+        @Nullable
+        String getImplementationClassName();
+
+        /**
+         * The input value property hashes.
+         * <p>
+         * key = property name
+         * <p>
+         * Ordered by key, lexicographically.
+         * No null keys or values.
+         * Never empty.
+         */
+        @Nullable
+        Map<String, byte[]> getInputValueHashesBytes();
+
+        /**
+         * Traverses the input properties that are file types (e.g. File, FileCollection, FileTree, List of File).
+         * <p>
+         * If there are no input file properties, visitor will not be called at all.
+         * <p>
+         * This is using the visitor from {@link SnapshotTaskInputsBuildOperationType} since there is no difference
+         * between tasks and transforms in this regard. Later we can unify the transform and the task build operation type.
+         */
+        void visitInputFileProperties(SnapshotTaskInputsBuildOperationType.Result.InputFilePropertyVisitor visitor);
```

Comment:
:x: This shouldn't refer to the visitor from `SnapshotTaskInputsBuildOperationType`. I'd duplicate the code for the visitor here (or extract a shared visitor class?) instead.

Reply:
We discussed that and we decided to not introduce a new type for the new specific `SnapshotTransformInputsBuildOperationType` that does exactly the same thing as `SnapshotTaskInputsBuildOperationType`. Instead we'd move both of these operations to a new type when we don't use this build operation anymore. I agree that it isn't nice to reference the `SnapshotTaskInputsBuildOperationType` visitor from here, though I think like this is better than to duplicate the old type and then also need to duplicate the logic for consuming the same things on the GE side.

Reply:
To me that doesn't feel like a good compromise. Even if we want to declare the visitor only once, it would be better to declare it in an external location. You could do something like `interface InputFilePropertyVisitor extends GeneralInputFilePropertyVisitor` in `SnapshotTaskInputsBuildOperationType.Result` for backwards compatibility.

Reply:
My understanding is that if we'd do what you suggest and compile the GE plugin against it, then it would try to use the methods on `GeneralInputFilePropertyVisitor` and fail on older Gradle versions. I think I'd need to do the extension the other way around (`interface GeneralInputFilePropertyVisitor extends InputFilePropertyVisitor`) and that doesn't seem good either. I can try your suggestion and see if the plugin would still work with older Gradle versions.

Reply:
I'm fairly sure it would work. When you compile against the more specific class, the method calls will reference the more specific type, even though the methods are declared in the more general type. This is something we need to deal with around bytecode upgrades, too (i.e. if you call `getEnabled()` on a `JavaCompile` object, you don't get `Task.getEnabled()` in the bytecode, or `AbstractTask.getEnabled()`, but `JavaCompile.getEnabled()`).

Reply:
I won't compile against the more specific type, since I want to re-use the code on the capturing side for both transforms and tasks. So I'd compile against the more general type and that will fail on old Gradle versions.

Reply:
I extracted a `FilePropertyVisitor` that is not dependent on tasks anymore.